src/magic_folder/config.py

```diff
@@ -0,0 +1,419 @@
+            must already exist.
+
+        :param FilePath state_path: the configuration and state
+            directory (which should not already exist)
+
+        :param LocalAuthor author: the signer of snapshots created in
+            this folder
+
+        :param unicode collective_dircap: the read-capability of the
+            directory defining the magic-folder.
+
+        :param unicode upload_dircap: the write-capability of the
+            directory we upload data into.
+
+        :param FilePath magic_directory: local path to the folder we
+            synchronize for this magic-folder.
+
+        :returns: a MagicFolderConfig instance
+        """
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute("SELECT name FROM magic_folders WHERE name=?", (name, ))
+            if len(cursor.fetchall()):
+                raise ValueError(
+                    "Already have a magic-folder named '{}'".format(name)
+                )
+        if not magic_path.exists():
+            raise ValueError(
+                "'{}' does not exist".format(magic_path.path)
+            )
+        if state_path.exists():
+            raise ValueError(
+                "'{}' already exists".format(state_path.path)
+            )
+
+        mkdir(state_path.path)
+        stash_path = state_path.child("stash")
+        mkdir(stash_path.path)
+        db_path = state_path.child("state.sqlite")
+        connection = sqlite3.connect(db_path.path)
+        with connection:
+            cursor = connection.cursor()
+            cursor.execute("BEGIN IMMEDIATE TRANSACTION")
+            cursor.executescript(_magicfolder_config_schema)
+            connection.commit()
+            cursor.execute(
+                "INSERT INTO version (version) VALUES (?)",
+                (_magicfolder_config_version, )
+            )
+            # default configuration values
+            cursor.execute(
+                "INSERT INTO CONFIG (author_name, author_private_key, stash_path, collective_dircap, upload_dircap, magic_directory, poll_interval) VALUES (?, ?, ?, ?, ?, ?, ?)",
+                (
+                    author.name,
+                    author.signing_key.encode(Base32Encoder),
+                    stash_path.path,
+                    collective_dircap,
+                    upload_dircap,
+                    magic_path.path,
+                    poll_interval,
+                )
+            )
+            cursor.execute(
+                "INSERT INTO stash (path) VALUES (?)",
+                (stash_path.path, )
+            )
+
+        config = MagicFolderConfig(
+            name=name,
+            database=connection,
+        )
+
+        # add to the global config
+        with self.database:
+            cursor = self.database.cursor()
+            cursor.execute(
+                "INSERT INTO magic_folders VALUES (?, ?)",
+                (name, state_path.path)
+            )
```

Comment:
This could fail if someone is doing something wacky - trying to create two magic-folders with the same name but different paths.  They'll race to this point and then one will succeed and the other will fail and leave intermediate state lying around.  Making the whole function atomic would probably address this.